### PR TITLE
Lw/story 3207/cancel

### PR DESCRIPTION
**Pull request by johnnieskywalker**
# [Title](https://clubhouseLink)

<please insert a clubhouse link above>
  
## Changes 👷‍♀️
  <Please add short list of changes>
- item 1
  
## How to test 🧪
  <Please explain how to test your changes>
- step 1 ...
    
## Definition of done ✔️

- [ ] Acceptance criteria for each issue met
- [ ] Unit tests written where needed and passing
- [ ] End-to-end tests for happy path
- [ ] Documentation updated <When applicable>
- [ ] Project builds without errors
- [ ] Non-functional requirements met
- [ ] Code reviewed and functionality tested by the reviewer (locally, Heroku, etc.)
- [ ] Feature verified and accepted by Product Owner
- [ ] Project deployed to production environment

### CODE REVIEW

**Comment by johnnieskywalker** [comment]
image

**Comment by johnnieskywalker** [comment]
image

**Comment by johnnieskywalker** [comment]
Is it working and actually cancelling SL on goerli ? https://goerli.etherscan.io/tx/0x13fa850e04bb7af4182f89428fe3f7ef80fa93875e6011b08e1e9142763c780f
looks like it added trigger I wanted to remove ? 
https://ethtx.info/goerli/0x13fa850e04bb7af4182f89428fe3f7ef80fa93875e6011b08e1e9142763c780f/

File: features/automation/controls/CancelSlFormControl.tsx [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        }
+
+        const sendTxErrorHandler = () => {
+          finishLoader(false)
+        }
+
+        const txData = prepareTriggerData(vaultData, collateralActive, selectedSLValue)
+
+        const waitForTx = txHelpers
+          .sendWithGasEstimation(addAutomationBotTrigger, txData)
```
Comment (adamskrodzki) [inline features/automation/controls/CancelSlFormControl.tsx:184]: remove here

File: features/automation/controls/CancelSlFormControl.tsx [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+          .subscribe(txSendSuccessHandler, sendTxErrorHandler)
+      },
+      isLoading: false,
+      isRetry: false,
+      disabled: isOwner,
+    }
+
+    const props: CancelSlFormLayoutProps = {
+      liquidationPrice: vaultData.liquidationPrice,
+      addTriggerConfig: addTriggerConfig,
```
Comment (adamskrodzki) [inline features/automation/controls/CancelSlFormControl.tsx:194]: remove probably

File: features/automation/controls/CancelSlFormLayout.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import React from 'react'
+
+import {
+  RetryableLoadingButton,
+  RetryableLoadingButtonProps,
+} from '../../../components/dumb/RetryableLoadingButton'
+
+export interface CancelSlFormLayoutProps {
+  liquidationPrice: BigNumber
+  addTriggerConfig: RetryableLoadingButtonProps
```
Comment (adamskrodzki) [inline features/automation/controls/CancelSlFormLayout.tsx:17]: remove

**Comment by johnnieskywalker** [comment]
https://goerli.etherscan.io/tx/0xbe0597493e6a16478d9b11686640c86cabbcdfd2b1f43659f20d3416b0afdcf8#eventlog it works

Code needs some refactoring, solving TODO's and making more common structures for adding and removing trigger 

in AutomationTriggersData this event could be handled on a frontend, may be also as next task

**Comment by johnnieskywalker** [comment]
Transactions done with this code right now :
adding:
https://goerli.etherscan.io/tx/0xf94a8adadf54016a6264ed034f0bd4e72875d21b81ac942e6bc35a77661e071a#eventlog
removing:
https://goerli.etherscan.io/tx/0xf27a80e9cd442a731f5a8a2c1c3d9387a0574edc85170eb7ec054a4e61898389#eventlog

File: blockchain/calls/automationBot.ts [hunk 4]
```diff
@@ -31,14 +39,36 @@ function getAddAutomationTriggerCallData(
...
   },
   prepareArgs: (data, context) => {
     const { automationBot } = context
-
-    console.log('Inside addAutomationBotTrigger.prepareArgs', automationBot.address)
-    console.log('Inside addAutomationBotTrigger.prepareArgs - data', data)
     return [automationBot.address, getAddAutomationTriggerCallData(data, context).encodeABI()]
   },
 }
+// TODO ŁW refactor use template method pattern and getAddAutomationTriggerCallData
```
Comment (johnnieskywalker) [inline blockchain/calls/automationBot.ts:49]: Maybe this one is not relevant anymore this types are slightly different and use different method which consists of 3 steps 1 defining context, 2 removing allowance, 3 calling contract

File: features/automation/controls/CancelSlFormLayout.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      <Text variant="strong">{summaryHeader}</Text>
+      <Grid columns={2}>
+        <Box>
+          <Text variant="subheader">{liquidation}</Text>
+        </Box>
+        <Box>{`$${formatCryptoBalance(props.liquidationPrice)}`}</Box>
+        <Box>
+          <Text variant="subheader">{total}</Text>
+        </Box>
+        <Box>$250.00</Box>
```
Comment (fenris85) [inline features/automation/controls/CancelSlFormLayout.tsx:46]: why hardcoded value?

File: features/automation/controls/CancelSlFormLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+export interface CancelSlFormLayoutProps {
+  liquidationPrice: BigNumber
+  removeTriggerConfig: RetryableLoadingButtonProps
+  txState?: TxState<AutomationBotRemoveTriggerData>
+}
+
+export function CancelSlFormLayout(props: CancelSlFormLayoutProps) {
+  const { t } = useTranslation()
+  const header = t(`cancel-stoploss.header`)
```
Comment (fenris85) [inline features/automation/controls/CancelSlFormLayout.tsx:23]: why are you assigning these translations to variables and not using them directly in returned component?

File: features/automation/common/StopLossTriggerDataExtractor.ts [hunk 7]
```diff
@@ -15,17 +19,19 @@ export interface StopLossTriggerData {
...
   isToCollateral: boolean
+  triggerId: number
 }
 
 export function extractSLData(data: TriggersData): StopLossTriggerData {
   const doesStopLossExist = data.triggers ? data.triggers.length > 0 : false
   if (doesStopLossExist) {
-    const slRecord: TriggerRecord | undefined = data.triggers ? last(data.triggers) : undefined
+    const slRecord = last(data.triggers)
     if (!slRecord) throw data /* TODO: This is logically unreachable, revrite so typecheck works */
```
Comment (johnnieskywalker) [inline features/automation/common/StopLossTriggerDataExtractor.ts:29]: could we assertNever using Exhaustive check?

File: features/automation/controls/CancelSlFormLayout.tsx (hunk 5 above)
Comment (johnnieskywalker) [reply features/automation/controls/CancelSlFormLayout.tsx:46]: :/ forgot about it sorry need to be calculated from transactions

File: features/automation/controls/CancelSlFormLayout.tsx (hunk 6 above)
Comment (johnnieskywalker) [reply features/automation/controls/CancelSlFormLayout.tsx:23]: Yeah there's no value added in it

**Review by adamskrodzki — APPROVED** [review]

**Comment by johnnieskywalker** [comment]
Decided with @adamskrodzki that PR can be merged to `feature/automation` however comments https://github.com/OasisDEX/oasis-borrow/pull/667#discussion_r773740102
https://github.com/OasisDEX/oasis-borrow/pull/667#discussion_r773743387
will be addressed

**Pull request merged**